src/mon/MonClient.cc

```diff
@@ -688,22 +695,47 @@ MonConnection& MonClient::_add_conn(unsigned rank, uint64_t global_id)
 
 void MonClient::_add_conns(uint64_t global_id)
 {
-  map<uint16_t, vector<unsigned>> rank_by_priority;
-  for (const auto& m : monmap.mon_info) {
-    rank_by_priority[m.second.priority].push_back(monmap.get_rank(m.first));
-  }
-  vector<unsigned> ranks;
-  ceph_assert(!rank_by_priority.empty());
-  ranks = rank_by_priority.begin()->second;
+  // collect the next batch of candidates who are listed right next to the ones
+  // already tried
+  auto get_next_batch = [this]() -> vector<unsigned> {
+    multimap<uint16_t, unsigned> ranks_by_priority;
+    boost::copy(monmap.mon_info | boost::adaptors::filtered([this](auto& info) {
```

Comment:
@cbodley could you help review this part? i think you are one of our experts of boost libraries. =)

the goal of this part is to collect the candidate monitors in `monmap.mon_info` into `rank`, `rank` will exclude the monitors in `this->tried`, and only use the monitors with lowest priority in this batch.

Reply:
this is neat! i don't have any experience with the boost/range library, but it looks a lot like the c++20 ranges library. and the code certainly looks like it will do what you describe - but if there's uncertainty, it's probably worth separating into a helper function and writing a unit test for it

Reply:
thank you @cbodley ! i will try to add one when porting this change to crimson's mon client.